### PR TITLE
feat: Implement `for-each` binding functionality

### PR DESCRIPTION
**Pull request by nicovillamonte**
Esta descripción del PR la generé con IA y me pareció muy completa así que la dejo tal cual:

---

# Feature: `for-each` Binding

## Overview

This PR implements the `for-each` binding directive for PelelaJS, enabling declarative iteration over collections with full reactive support, nested scope access, and efficient DOM updates.

## Syntax

```html
<tr for-each="user of users">
  <td bind-value="user.id"/>
  <td bind-value="user.name"/>
  <td bind-value="user.email"/>
</tr>
```

### 1. Declarative Iteration
- Simple, intuitive syntax: `for-each="item of collection"`
- Works with any HTML element
- Supports nested elements and complex structures

### 2. Nested Scope Support
Each replicated element has access to:
- **Loop variable**: Direct access to the current item (e.g., `user`, `item`)
- **Nested properties**: Access to item properties using dot notation (e.g., `user.profile.email`)
- **Parent ViewModel**: Access to all parent ViewModel properties

```html
<div for-each="item of items">
  <span bind-value="parentProperty"></span>  <!-- Parent ViewModel -->
  <span bind-value="item.text"></span>       <!-- Loop item property -->
</div>
```

### 3. Multiple Bindings on Same Element
The `for-each` directive can be combined with other bindings on the same element:

```html
<!-- Conditional rendering with for-each -->
<li for-each="item of items" if="item.visible">
  <span bind-value="item.text"></span>
</li>

<!-- bind-value with for-each (useful for select options) -->
<select bind-value="selected">
  <option for-each="option of options" bind-value="option.label"></option>
</select>

<!-- Multiple bindings -->
<span for-each="item of items" 
      if="item.active" 
      bind-value="item.text" 
      bind-class="item.className"></span>
```

### 4. Reactive Updates
Array mutations are automatically detected and trigger efficient DOM updates:
- **Intercepted methods**: `push`, `pop`, `shift`, `unshift`, `splice`, `sort`, `reverse`
- **Efficient rendering**: Only adds/removes necessary elements instead of re-rendering everything
- **Smart updates**: Compares current vs previous array length and updates accordingly

```typescript
// In your ViewModel
addUser() {
  this.users.push({ id: 4, name: "Dave", email: "dave@example.com" });
  // DOM automatically updates with new <tr>
}
```

#### 1. Enhanced Reactive Proxy ([`reactiveProxy.ts`](packages/core/src/reactivity/reactiveProxy.ts))
Extended the reactive system to intercept array mutation methods:
- Wraps array methods to trigger `onChange` callback
- Applied automatically to all arrays in the ViewModel
- Works with both initial arrays and arrays assigned later

#### 2. ForEach Binding ([`bindForEach.ts`](packages/core/src/bindings/bindForEach.ts))
New binding system that handles:
- **Expression parsing**: Validates `"item of collection"` syntax
- **Template cloning**: Preserves original element as template
- **Extended ViewModel**: Creates scoped proxy for each replicated element
- **Nested bindings**: Processes all bindings within replicated elements
- **Efficient updates**: Tracks previous array length and only updates differences

Key functions:
- `setupForEachBindings()`: Finds and configures for-each elements
- `renderForEachBindings()`: Manages DOM updates when collections change
- `createExtendedViewModel()`: Creates scoped proxy with access to loop variable and parent

#### 3. Updated Types ([`types.ts`](packages/core/src/bindings/types.ts))
- Added `ForEachBinding` type
- Updated `BindingsCollection` to include `forEachBindings`

#### 4. Integration ([`setupBindings.ts`](packages/core/src/bindings/setupBindings.ts))
- **Critical**: `for-each` bindings are processed **BEFORE** other bindings
- This ensures replicated elements exist in the DOM before their bindings are processed
- Rendering order: for-each → value → if → class → style

#### 5. HTMLSelectElement Support ([`bindValue.ts`](packages/core/src/bindings/bindValue.ts))
Enhanced `bind-value` to properly handle `<select>` elements:
- Treats `HTMLSelectElement` as input element
- Sets `value` property instead of `textContent` (preserves children)
- Uses `change` event instead of `input` for proper select behavior
- Uses `Reflect.set` for type-safe property assignment

#### 1. Autocomplete ([`extension.js`](tools/pelela-vscode/extension.js))
- Added `for-each` to attribute suggestions
- Provides snippet: `for-each="${1:item} of ${2:collection}"`
- Suggests ViewModel properties when typing inside `for-each` value

#### 2. Go to Definition ([`extension.js`](tools/pelela-vscode/extension.js))
Intelligent navigation for complex expressions:

**For `for-each="item of collection"`:**
- Ctrl+Click on `collection` → Goes to ViewModel property definition

**For nested properties like `item.text`:**
- Ctrl+Click on `item` → Goes to loop variable in `for-each`
- Ctrl+Click on `text` → Searches for property in ViewModel

**Features:**
- Works across multiple lines
- Works at any nesting level
- Finds closest `for-each` in parent hierarchy
- Supports optional properties (`active?:`)

#### 3. Syntax Highlighting ([`pelela.tmLanguage.json`](tools/pelela-vscode/syntaxes/pelela.tmLanguage.json))
Enhanced syntax highlighting for `for-each` values:
- Loop variable (`item`) → Variable color
- Keyword (`of`) → Keyword color
- Collection (`items`) → Property color

Makes it visually similar to JavaScript's `for...of` syntax.

#### 4. Custom Data ([`html-custom-data.json`](tools/pelela-vscode/html-custom-data.json))
- Added `for-each` to global attributes
- Provides description and documentation

### Comprehensive Test Suite ([`bindForEach.test.ts`](packages/core/src/bindings/bindForEach.test.ts))
20 tests covering:
- Expression parsing and validation
- Initial rendering with various HTML structures
- Dynamic array manipulation (push, pop, splice)
- Nested property access (`user.profile.email`)
- Multiple bindings on same element
- Nested bindings support (if, bind-class, etc.)
- Parent ViewModel property access
- Edge cases (empty arrays, invalid expressions)
- Select/option elements
- Table rows
- List items

**Result**: 121 total tests passing (20 for for-each, 101 existing)

### Basic List
```html
<ul>
  <li for-each="item of items">
    <span bind-value="item.text"></span>
  </li>
</ul>
```

### Conditional Items
```html
<ul>
  <li for-each="item of items" if="item.visible">
    <span bind-value="item.text"></span>
  </li>
</ul>
```

### Table with Data
```html
<table>
  <tbody>
    <tr for-each="user of users">
      <td bind-value="user.id"/>
      <td bind-value="user.name"/>
      <td bind-value="user.email"/>
    </tr>
  </tbody>
</table>
```

### Select with Options
```html
<select bind-value="selected">
  <option for-each="option of options" bind-value="option.label"></option>
</select>
```

### Nested Properties
```html
<div for-each="user of users">
  <h3 bind-value="user.profile.name"></h3>
  <p bind-value="user.profile.bio"></p>
</div>
```

## Performance

- **Efficient updates**: Only adds/removes changed elements
- **No full re-renders**: Existing elements are updated in place
- **Minimal overhead**: Uses native Proxy API for reactivity
- **Smart diffing**: Compares array lengths to determine update strategy

## Breaking Changes

None. This is a new feature that doesn't affect existing functionality.

## Migration Guide

No migration needed. Simply start using `for-each` in your templates:

```html
<!-- Before: Manual duplication -->
<div><span bind-value="item1"></span></div>
<div><span bind-value="item2"></span></div>
<div><span bind-value="item3"></span></div>

<!-- After: Dynamic iteration -->
<div for-each="item of items">
  <span bind-value="item.value"></span>
</div>
```

### Why for-each is processed first
The `for-each` binding must be processed before other bindings because:
1. It creates new DOM elements dynamically
2. Other bindings need these elements to exist
3. Processing order: `for-each` → `value` → `if` → `class` → `style` → `click`

### Extended ViewModel Implementation
Uses a Proxy to provide nested scope:
- Checks if property starts with loop variable name
- Falls back to parent ViewModel for other properties
- Supports deep property access (e.g., `item.nested.property`)
- Uses reference object (`itemRef`) to allow updates without breaking bindings

### Array Reactivity
Wraps array mutation methods at three levels:
1. **Initial arrays**: Wrapped when ViewModel is created
2. **New arrays**: Wrapped when assigned to a property
3. **Accessed arrays**: Wrapped when accessed via Proxy getter

This ensures all array mutations trigger re-renders regardless of when/how the array was created.

### Core Package
- `packages/core/src/reactivity/reactiveProxy.ts` - Array reactivity
- `packages/core/src/bindings/bindForEach.ts` - ForEach binding (NEW)
- `packages/core/src/bindings/bindForEach.test.ts` - Tests (NEW)
- `packages/core/src/bindings/types.ts` - Type definitions
- `packages/core/src/bindings/setupBindings.ts` - Integration
- `packages/core/src/bindings/bindValue.ts` - HTMLSelectElement support
- `packages/core/src/bootstrap/bootstrap.ts` - Type fix

### VSCode Extension
- `tools/pelela-vscode/extension.js` - Autocomplete & Go to Definition
- `tools/pelela-vscode/syntaxes/pelela.tmLanguage.json` - Syntax highlighting
- `tools/pelela-vscode/html-custom-data.json` - Custom data

## Future Enhancements

Potential improvements for future PRs:
- Index variable support: `for-each="(item, index) of items"`
- Key-based reconciliation for better performance with reordering
- Support for iterating over objects: `for-each="[key, value] of object"`
- `bind-items` directive for simpler select/option scenarios (as discussed in notes)

### CODE REVIEW

File: packages/core/src/bindings/bindForEach.test.ts [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        <ul>
+          <li for-each="item of items">Item</li>
+        </ul>
+      `;
+
+      const viewModel = { items: [] };
+      setupForEachBindings(container, viewModel);
+
+      const li = container.querySelector("li");
+      expect(li).toBeNull();
```
Comment (fdodino) [inline packages/core/src/bindings/bindForEach.test.ts]: ah, este test me mató, pero confío en que la implementación es la correcta

File: packages/core/src/bindings/bindForEach.test.ts [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+      const bindings = setupForEachBindings(container, viewModel);
+
+      expect(bindings).toHaveLength(0);
+      expect(consoleSpy).toHaveBeenCalled();
+      consoleSpy.mockRestore();
+    });
+
+    it("should handle non-array collection", () => {
+      container.innerHTML = '<div for-each="item of notArray"></div>';
+      const viewModel = { notArray: "string" };
```
Comment (fdodino) [inline packages/core/src/bindings/bindForEach.test.ts]: 👏🏼

File: packages/core/src/bindings/bindForEach.ts [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+  };
+}
+
+function mapElementPath(
+  sourceElement: HTMLElement,
+  sourceRoot: HTMLElement,
+  targetRoot: HTMLElement,
+): HTMLElement {
+  if (sourceElement === sourceRoot) {
+    console.log(
```
Comment (fdodino) [inline packages/core/src/bindings/bindForEach.ts]: uh todo este log me parece buenísimo, podemos elegir contarlo cuando se sientan más cancheros

File: packages/core/src/bindings/bindForEach.ts [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+export function renderForEachBindings<T extends object>(
+  bindings: ForEachBinding[],
+  viewModel: ViewModel<T>,
+): void {
+  for (const binding of bindings) {
+    const collection = viewModel[binding.collectionName];
+
+    if (!Array.isArray(collection)) {
+      continue;
```
Comment (fdodino) [inline packages/core/src/bindings/bindForEach.ts]: entiendo que vos lo atajaste en el `setupForEachBindings` por eso no tirás error, no?

File: packages/core/src/bindings/setupBindings.ts [hunk 5]
```diff
@@ -19,6 +21,7 @@ export function setupBindings<T extends object>(
   setupClickBindings(root, viewModel);
 
   const render = () => {
+    renderForEachBindings(bindings.forEachBindings, viewModel);
```
Comment (fdodino) [inline packages/core/src/bindings/setupBindings.ts]: no se si lo pusiste en algún otro lado, agregaría un comentario explicando que el orden es importante (por si a alguien se le ocurre ordenarlo alfabéticamente, jajaj)

**Review by fdodino — APPROVED** [review]
Una joya, un rubí.

Me quedé pensando en que la documentación que generaste con la IA quizás convenga crearla dentro de este repo nomás, armando unas github pages y después publicando en una URL onda http://pelela.uqbar-project.org como hicimos en su momento con Arena.

La verdad, un golazo, tenía algunos comentarios de hincha pero que podemos pelotear a futuro:
- extraer cada iteración (cada for/while de las funciones setup y render) en métodos/funciones aparte, quedaron un poco largas las funciones y calculo que tienen clases de equivalencia diferente en los tests, así que podríamos refactorizarlas aparte si en algún momento nos pinta
- en el ejemplo de binding de tabla y de select, yo buscaría a ver si se puede unificar el dominio, para que no sean de cosas distintas. Por ahora dejémoslo así, yo después pienso algo con tiempo.

File: packages/core/src/bindings/setupBindings.ts (hunk 5 above)
Comment (nicovio) [reply packages/core/src/bindings/setupBindings.ts]: sería medio imprudente porque es cambiar el orden de ejecución de una función. Pero estoy con vos, agregaría el comentario si o si.

File: examples/for-each/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    { descripcion: "Segundo Puesto", active: true },
+    { descripcion: "Tercero" },
+  ];
+
+  tipoApuesta: string = this.tiposApuesta[0].descripcion;
+  newUserName: string = "";
+  newUserEmail: string = "";
+
+  addUser() {
+    if (!this.newUserName || !this.newUserEmail) return;
```
[the remark this answers is not in the document]
Comment (nicovillamonte) [reply examples/for-each/src/app.ts]: Si, en realidad este no es un ejemplo tal cual para dar clases, es un ejemplo para mostrar rapidamente todos los usos del foreach y poder probarlo tambien nosotros mientras desarrollamos, despues si queres agregamos o mejoramos este mismo ejemplo pero no lo pense como para dar clases con este, mas que nada necesitaba probar que todo ande y decidí subirlo para que también lo puedan ver ustedes, pero en principio lo tenía dentro de un ejemplo local que hice dentro de una carpeta `local/` que esta agregada al `.gitignore` y no se sube.

[the remark this answers is not in the document]
Comment (fdodino) [reply examples/for-each/src/app.ts]: sí, tenés razón. Recapitulando me di cuenta de que necesitamos tener ejemplos como los de Angular, React, Svelte, y tienen que estar en repos separados.

File: packages/core/src/bindings/bindForEach.ts (hunk 3 above)
Comment (nicovillamonte) [reply packages/core/src/bindings/bindForEach.ts]: Esos logs son obra de la IA jajajja, hace lindos logs, pero no quería subirlo con todo eso, por eso abrí el issue de los logs, por ahora los dejo pero despues con ese issue lo mejoramos para que no haya millones de logs por cada cosa que se haga jajaj.

**Comment by nicovio** [comment]
[acá](https://github.com/uqbar-project/pelelajs/pull/26#issuecomment-3597682190) te comenté lo que me parece que hace fallar al CI @nicovillamonte

**Comment by nicovio** [comment]
Revisé los últimos cambios y tienen buena pinta:

- Lo del `DependencyTracker` está muy bueno para hacer re-renders más inteligentes y solo actualizar los bindings que realmente se ven afectados por cada cambio. 
- Lo de hacer reactivas las propiedades nesteadas también era importante.
- Vi que refactorizaste creando varios helpers reutilizables, banco.

Se me pueden escapar detalles de la implementación pero me parecen cambios positivos. Lo podemos mergear y si en el futuro vemos algo para corregir creamos nuevos issues.
